### PR TITLE
SNOW-2266770: String datatype support: gherkins and ODBC(without temporal and interval conversion full support or support for C locale)

### PR DESCRIPTION
**Pull request by sfc-gh-jszczerbinski**

### CODE REVIEW

**Comment by sfc-gh-makowalski** [comment]
When running `cargo test` on this PR's branch, some unit tests fail but I don't see this issue on the CI. The tests in question:
`test api::data::tests::read_to_char::truncates_when_buffer_too_small ... FAILED`
`test api::data::tests::unsupported_types::returns_error_for_wchar_target_type ... FAILED`
Is there something wrong with my setup or with these unit tests?

**Comment by sfc-gh-makowalski** [comment]
I think we need to update the tests around truncation, based on the old ODBC driver's tests (and some microsoft doc) we always want to include the null terminator at the end of the string, even a truncated one, because applications rely on strlen to stop at null terminator.
Let me know if you want me to push the fixes I have ready.

**Comment by sfc-gh-makowalski** [comment]
Also there is one test that I think you omitted, namely `returns_error_on_wchar_target_type()` in `odbc/src/api/data.rs`.

File: .cursor/rules/datatype_gherkins.mdc [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+Focus on following types of tests
+- “Happy path” tests - simply select several hardcoded values, easy to debug.
+- Corner cases - “weird and special” values (min, max, NaN) or specific wrapper behaviours.
+- Generative tests - execute multiple queries on randomized data. This will allow testing on multiple different values. The seed used for generation needs to be logged so it will be possible to re-create the error.
+
+# Test scenarios
+- Simple selects - Selecting literals, selecting from table should be covered by all test kinds (6 tests in total). Generative tests should execute simple operations multiple times on different values.
+- Simple insert with a binding - Insert value to a table and then select it back expecting the same value. (2 tests, simple and generic).
+- Multiple chunks downloading - this should be a single test, which generates large but reasonable amount of data (~10^6 values - to ensure they’re downloaded in at least two chunks), inserts all to the Snowflake table using the driver, and then executing SELECT * from that table. The result should be equal to the original data.
```
Comment (sfc-gh-pczajka) [inline .cursor/rules/datatype_gherkins.mdc:17]:
```suggestion
- Multiple chunks downloading - this should be a single test, which generates large but reasonable amount of data (~10^6 values - to ensure they’re downloaded in at least two chunks), inserts all to the Snowflake table using the driver, and then executing SELECT * from that table. The result should be equal to the original data. Make sure to use ORDER BY in SQL queries.
```
Snowflake does not guarantee order of the result - recent python finding :)

File: .cursor/rules/datatype_gherkins.mdc [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+description: Guidelines for generating datatype tests
+alwaysApply: true
+---
+
+# Test kinds
+
+Focus on following types of tests
+- “Happy path” tests - simply select several hardcoded values, easy to debug.
+- Corner cases - “weird and special” values (min, max, NaN) or specific wrapper behaviours.
+- Generative tests - execute multiple queries on randomized data. This will allow testing on multiple different values. The seed used for generation needs to be logged so it will be possible to re-create the error.
```
Comment (sfc-gh-pczajka) [inline .cursor/rules/datatype_gherkins.mdc]: I think this is outdated

File: .cursor/rules/datatype_gherkins.mdc [hunk 3]
```diff
@@ -0,0 +1,24 @@
+---
+description: Guidelines for generating datatype tests
+alwaysApply: true
```
Comment (sfc-gh-pczajka) [inline .cursor/rules/datatype_gherkins.mdc]: why true?

File: .cursor/rules/datatype_gherkins.mdc [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+# Test scenarios
+- Simple selects - Selecting literals, selecting from table should be covered by all test kinds (6 tests in total). Generative tests should execute simple operations multiple times on different values.
+- Simple insert with a binding - Insert value to a table and then select it back expecting the same value. (2 tests, simple and generic).
+- Multiple chunks downloading - this should be a single test, which generates large but reasonable amount of data (~10^6 values - to ensure they’re downloaded in at least two chunks), inserts all to the Snowflake table using the driver, and then executing SELECT * from that table. The result should be equal to the original data.
+- Type casting (if applicable, for example ODBC skips this, as the user chooses which type data should be cast to) - simple select which ensures that data are being cast to appropriate types. This should be the only test which documents type conversion (not to duplicate complicated comments).
+
+# General
+- Please refer to online snowflake docs on details about how the datatypes function in Snowflake
```
Comment (sfc-gh-pczajka) [inline .cursor/rules/datatype_gherkins.mdc]: I'd add "refer to existing types/*.feature files as an example

File: .cursor/rules/datatype_gherkins.mdc [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+# Steps
```
Comment (sfc-gh-pczajka) [inline .cursor/rules/datatype_gherkins.mdc:24]: FYI - I've created an .md file with LLM analysys of documentation and other driver tests - it was useful for generating corner case values

File: tests/definitions/shared/types/string.feature [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    Then the result should contain:
+      | str1  | str2        | str3                  |
+      | hello | Hello World | Snowflake Driver Test |
+
+  @odbc_e2e
+  Scenario: should select string literals with corner case values
+    # Corner cases: empty string, single character, whitespace-only, unicode characters, escape sequences
+    Given Snowflake client is logged in
+    When Query selecting corner case string literals is executed
+    # Corner cases include:
```
Comment (sfc-gh-pczajka) [inline tests/definitions/shared/types/string.feature:26]: Nice!

File: tests/definitions/shared/types/string.feature [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    And The table is populated with string values
+    When Query "SELECT * FROM {table}" is executed
+    Then the result should contain the inserted hardcoded string values
+
+  @odbc_e2e
+  Scenario: should select corner case string values from table
+    Given Snowflake client is logged in
+    And A temporary table with VARCHAR column is created
+    And The table is populated with corner case string values
+    # Corner cases: empty string, max length string, unicode, special characters
```
Comment (sfc-gh-pczajka) [inline tests/definitions/shared/types/string.feature]: Those differ from "literals" test - why?

File: tests/definitions/shared/types/string.feature [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    When Query "SELECT * FROM {table}" is executed
+    Then the result should contain the inserted corner case string values
+
+  @odbc_e2e
+  Scenario: should select generative random string values from table
```
Comment (sfc-gh-pczajka) [inline tests/definitions/shared/types/string.feature]: We resigned from generative tests in this scope

File: tests/definitions/shared/types/string.feature [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    And The table is populated with 100 randomly generated string values
+    When Query "SELECT * FROM {table} ORDER BY id" is executed
+    Then all returned string values should match the generated values in order
+
+  # ============================================================================
+  # SIMPLE INSERT WITH BINDING (Simple, Generative)
+  # ============================================================================
+
+  @odbc_e2e
+  Scenario: should insert and select back hardcoded string values using parameter binding
```
Comment (sfc-gh-pczajka) [inline tests/definitions/shared/types/string.feature:79]: Could you add 'SELECT ?' case too?
The testing doc was updated with more details to help keep track of all cases:
https://docs.google.com/document/d/1FCioVbI1rGEQSsDRJQE-I_ob_8IjpU9-hyIZzkDM42w/edit?tab=t.0

File: .cursor/rules/datatype_gherkins.mdc (hunk 3 above)
Comment (sfc-gh-jszczerbinski) [reply .cursor/rules/datatype_gherkins.mdc]: Should be false, true

File: .cursor/rules/datatype_gherkins.mdc (hunk 2 above)
Comment (sfc-gh-jszczerbinski) [reply .cursor/rules/datatype_gherkins.mdc]: Removed

File: tests/definitions/shared/types/string.feature (hunk 9 above)
Comment (sfc-gh-jszczerbinski) [reply tests/definitions/shared/types/string.feature:79]: I will modify the cursor rule and add the test too.

**Review by sfc-gh-makowalski — CHANGES_REQUESTED** [review]
I don't see anything wrong with this PR except for some gaps between this and the old odbc driver. I'm willing to approve this if we agree on the list of the items that are missing so we for sure don't forget anything. @sfc-gh-jszczerbinski what do you think?

File: tests/definitions/shared/types/string.feature (hunk 7 above)
Comment (sfc-gh-jszczerbinski) [reply tests/definitions/shared/types/string.feature]: Now they are the same

File: tests/definitions/shared/types/string.feature (hunk 8 above)
Comment (sfc-gh-jszczerbinski) [reply tests/definitions/shared/types/string.feature]: Removed

File: .cursor/rules/datatype_gherkins.mdc (hunk 4 above)
Comment (sfc-gh-jszczerbinski) [reply .cursor/rules/datatype_gherkins.mdc]: Done

File: tests/definitions/shared/types/string.feature [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+    When String value 'Test binding value 日本語' is inserted using parameter binding
+    And Query "SELECT * FROM {table}" is executed
+    Then the result should contain the bound string value 'Test binding value 日本語'
+
+  
+  @odbc_e2e
+  Scenario: should select string literals using parameter binding
+    # SELECT binding test: Uses SELECT ?::VARCHAR to bind string values
+    Given Snowflake client is logged in
+    When Query "SELECT ?::VARCHAR, ?::VARCHAR, ?::VARCHAR" is executed with bound string values ['hello', 'Hello World', '日本語テスト']
```
Comment (sfc-gh-pczajka) [inline tests/definitions/shared/types/string.feature:91]:
```suggestion
    When Query "SELECT ?::<type>, ?::<type>, ?::<type>" is executed with bound string values ['hello', 'Hello World', '日本語テスト']
```

In case we want to parametrize test with all string synonyms; not blocking of does not make sense from your perspective

File: tests/definitions/shared/types/string.feature [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+    #   - Surrogate pair: '\U0001D11E' (𝄞 musical G clef)
+    Then the result should match the bound corner case value
+
+
+  # ============================================================================
+  # MULTIPLE CHUNKS DOWNLOADING
+  # ============================================================================
+
+  @odbc_e2e
+  Scenario: should download string data in multiple chunks
```
Comment (sfc-gh-pczajka) [inline tests/definitions/shared/types/string.feature:121]: WDYTH about adding test with large table? It is present in integer tests (for example)

**Review by sfc-gh-makowalski — APPROVED** [review]

File: tests/definitions/shared/types/string_lob.feature [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+    And Query "SELECT val, LENGTH(val) as len FROM {table}" is executed
+    Then the result should show length 16777216
+    And the returned string should exactly match the generated string
+
+  @odbc_e2e
+  Scenario: should handle LOB string at maximum 128 MB limit with increased LOB size
+    # Corner case: string at maximum LOB limit (128 MB) - requires Increased LOB Size feature
+    Given Snowflake client is logged in
+    And A temporary table with VARCHAR column is created
+    When A string of 134217728 random ASCII characters is generated and inserted
```
Comment (sfc-gh-pczajka) [inline tests/definitions/shared/types/string_lob.feature:22]:
```suggestion
    When A string of 134217728 ASCII characters is generated and inserted
```

or let's specify that hte seed is hardcoded (to avoid uncertainieties pointed out dufring generative tests discussion)

File: tests/definitions/shared/types/string_lob.feature [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+  # Snowflake LOB feature supports large VARCHAR values:
+  #   - Historical limit: 16 MB (16,777,216 bytes) per value
+  #   - Increased LOB Size feature: up to 128 MB (134,217,728 bytes) per value
+
+  @odbc_e2e
+  Scenario: should handle LOB string at historical 16 MB limit
+    # Corner case: string at the historical LOB limit (16 MB = 16,777,216 bytes)
+    Given Snowflake client is logged in
+    And A temporary table with VARCHAR column is created
+    When A string of 16777216 random ASCII characters is generated and inserted
```
Comment (sfc-gh-pczajka) [inline tests/definitions/shared/types/string_lob.feature:12]:
```suggestion
    When A string of 16777216 ASCII characters is generated and inserted
```

Same

File: .cursor/rules/datatype_gherkins.mdc [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+---
+
+# Test kinds
+
+Focus on following types of tests
+- “Happy path” tests - simply select several hardcoded values, easy to debug.
+- Corner cases - “weird and special” values (min, max, NaN) or specific wrapper behaviours.
+
+# Test scenarios
+- Simple selects - Selecting literals, selecting from table should be covered by all test kinds (6 tests in total). Generative tests should execute simple operations multiple times on different values.
```
Comment (sfc-gh-pczajka) [inline .cursor/rules/datatype_gherkins.mdc:13]:
```suggestion
- Simple selects -Selecting literals, selecting from table, both happy and corner cases (4 tests in total)
```

Updated from doc

File: tests/definitions/shared/types/string.feature (hunk 11 above)
Comment (sfc-gh-jszczerbinski) [reply tests/definitions/shared/types/string.feature:121]: With our current bind api support, those tests take forever to run. For me this is fine, I think we will have those tests in performance testing anyway. So I am ok with skipping those
